boa_engine/src/string/mod.rs

```diff
@@ -260,8 +362,45 @@ impl JsString {
                 // `allocate_inner` must return a valid pointer to newly allocated memory, meaning
                 // `ptr` and all `string`s should never overlap.
                 unsafe {
-                    ptr::copy_nonoverlapping(string.as_ptr(), data, count);
-                    data = data.add(count);
+                    match (ascii, string.variant()) {
+                        (true, JsStringSliceVariant::U8Ascii(s)) => {
+                            ptr::copy_nonoverlapping(s.as_ptr(), data.cast::<u8>(), count);
+                            data = data.cast::<u8>().add(count).cast::<u8>();
+                        }
+                        (true, JsStringSliceVariant::U16Ascii(s)) => {
+                            for (i, byte) in s.iter().copied().enumerate() {
+                                *data.cast::<u8>().add(i) = (byte & 0xFF) as u8;
+                            }
+                            data = data.cast::<u8>().add(count).cast::<u8>();
+                        }
+                        (false, JsStringSliceVariant::U8Ascii(s)) => {
+                            for (i, byte) in s.bytes().enumerate() {
+                                *data.cast::<u16>().add(i) = u16::from(byte);
+                            }
+                            data = data.cast::<u16>().add(count).cast::<u8>();
+                        }
+                        (false, JsStringSliceVariant::U8NonAscii(s, _)) => {
+                            for (i, byte) in s.encode_utf16().enumerate() {
+                                *data.cast::<u16>().add(i) = byte;
+                            }
+                            data = data.cast::<u16>().add(count).cast::<u8>();
+                        }
+                        (
+                            false,
+                            JsStringSliceVariant::U16Ascii(s)
+                            | JsStringSliceVariant::U16NonAscii(s),
+                        ) => {
+                            ptr::copy_nonoverlapping(s.as_ptr(), data.cast::<u16>(), count);
+                            data = data.cast::<u16>().add(count).cast::<u8>();
+                        }
+                        (
+                            true,
+                            JsStringSliceVariant::U8NonAscii(..)
+                            | JsStringSliceVariant::U16NonAscii(_),
+                        ) => {
+                            unreachable!()
+                        }
+                    }
```

Comment:
I think trying to recover the "asciiness" of an UTF-16 string gets too unwieldy, making it difficult to follow implementations like this. In my opinion, we should just offer two variants: ASCII and UTF-16, and if a string becomes UTF-16, it should just stay that way from that point. 

This also preemptively converts the strings to UTF-16, which I think is a plus: if the user converted some strings to UTF-16, they are probably using non-ASCII characters in some part of the code, which means they'll probably keep using UTF-16 chars in the future. Converting back to UTF-8 just does unnecessary work in this case.

Reply:
The problem is that if we do not keep the asciiness, then it makes access such as comparing hard (instead of just doing `JsStr == JsStr`) also hashing and any other operation, which are more common.

Reply:
> Yeah, that's the downside of this other approach. I think having to convert from ASCII strings to UTF-16 strings repeatedly, and traverse UTF-16 strings several times to assert that they're ASCII (e.g. trim_start, which remains cheap with the alternative approach) would be more costly than having to compare by u32 iterator.

Yes, about this I had some optimizations in mind about this, one approach was to was that we only check if what we trimmed was non-ascii (it's uncommon to have non-acsii whitespace).

Another approach I had was to return a type that can be non-ascii for ascii string and lazy evaluate to a known ascii/non-ascii type.

Both optimizations are valid and both can bee applied :)

> In addition to that, the comparison problem can be solved with a JsStr that has the UTF-16 representation and (optionally) the ASCII representation of the string, and we can offer a js_str! macro to convert from &str to JsStr.

If we concatenate two (`u8`) ascii strings it will allocate it in a `u16` array, defeating the purpose of this optimization.

> I just have a question. Why would hashing be affected?

Because hashing two strings that are both ascii and same characters but different types (`u8` and `u16`) (gives two different hashes)[https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b5df58a7e685f3f75133006727125ee3].

We would have to iterate though the ascii string and convert to `u16` and hash that.

Reply:
Ok, if we're going with this approach, is there any way to make `JsStringSlice` internal to the implementation?

My main worry is to pollute our API with 4 types of string references with slightly different representations and semantics.

Reply:
> My main worry is to pollute our API with 4 types of string references with slightly different representations and semantics.

We can put in under `#[doc(hidden)]`, it's a common way to hide implementation detail, we can't make it `pub(crate)` because `js_string!()` macro uses it.

It was getting a bit messy, but still better than all the [string v8 offers](https://github.com/danbev/learning-v8/blob/master/notes/string.md) :laughing:

Reply:
In the future we may want to implement ConsStrings, which is a string that keeps a pointer to two other strings, which would make string concatenation cheep for big strings.

Reply:
> We can put in under #[doc(hidden)], it's a common way to hide implementation detail, we can't make it pub(crate) because js_string!() macro uses it.

Then there's a way to make those private! We can rename the `utf16` macro to `js_str`, and the macro should convert an `&str` into its correct variant depending on whether it has non-ascii chars or not. I think that should remove the need to use `JsSliceString` on `js_string`.

Reply:
Just realized that this also has the advantage of not having to traverse a `str` at runtime to know if it has ascii chars or not :)

Reply:
> In the future we may want to implement ConsStrings, which is a string that keeps a pointer to two other strings, which would make string concatenation cheep for big strings.

That and inline strings, since it's pretty useless to allocate for strings of length less than 10-20 chars.